# PRACA DOMOWA 2

Plik *data.csv* zawiera dane medyczne Indian z plemienia Pima. Ostatnia kolumna (*Outcome*) zawiera binarną zmienną odpowiedzi, gdzie 1 oznacza, że dana osoba ma cukrzyce.

In [14]:
import pandas as pd
import numpy as np
import sklearn

data = pd.read_csv("data.csv")
data.head()

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148           72.0             35        0  33.6   
1            1       85            NaN             29        0   NaN   
2            8      183           64.0              0        0   NaN   
3            1       89           66.0             23       94   NaN   
4            0      137            NaN             35      168  43.1   

   DiabetesPedigreeFunction   Age  Outcome  
0                     0.627  50.0        1  
1                     0.351  31.0        0  
2                     0.672   NaN        1  
3                     0.167   NaN        0  
4                     2.288  33.0        1

### Zadanie 0 Podziel zbiór na testowy i treningowy w proporcjach 0.3 / 0.7

## 1. Preprocessing

Wczytany zbiór danych posiada braki danych. Większość klasyfikatorów, nie jest w stanie operować na tak wybrakowanych danych.  

  
*uwaga*: drzewa decyzyjne (używane później) potrafią obsługiwać braki danych za pomocą *podziałów zastępczych*

### Zadanie 1.1


Wypełnij brakujące dane  
a) średnią  
b) medianą  
c) modą   
\*d) przy pomocy metody Multivariate feature imputation z biblioteki *sklearn*  

Wyświetl 5 pierwszych wierszy przy pomocy metody *.head()*

In [ ]:
#a)



In [3]:
#b)



In [4]:
#c)



In [17]:
#d)
from sklearn.experimental import enable_iterative_imputer 
from sklearn.impute import IterativeImputer


## Ocena jakości modelu

### Zadanie 2.1

Podziel zbior na testowy i treningowy. Wytrenuj drzewo decyzyjne na zbiorze z punktu a) zadania 1.1 (ustaw wartość 'max_depth' równą 5). Wyrysuj (funkcja *plot_tree* oraz *plt.show()*) otrzymane drzewo korzystając z kodu w następnej komórce podstawiając swoje zmienne. 

In [6]:
from sklearn import tree

In [8]:
import matplotlib.pyplot as plt
plt.figure(figsize=(32, 18)) # tworzymy wiekszy obrazek, dostosuj figsize do swojej rozdzielczosci



### Zadanie 2.2

Dla stworzonego drzewa oblicz  
a) Accuracy  
b) Error Rate  
c) Precision, Recall (True Positive Rate)
d) False Positive Rate


Wyrzuc ze zbioru testowego 90% wierszy z klasą 1. Porównaj Recall z Accuracy.  

Czy Accuracy jest dobrą metryką dla niezrównoważonych danych?  
Która z tych metryk jest najważniejsza w przypadku przewidywania czy dana osoba jest chora?

### Zadanie 2.3

Na zbiorze z Zad 1.1a Zmierz Accuracy dla drzew wytrenowanych z wartościami parametru 'max_depth' 3, 5, 10 oraz 15.

Zmierz Accuracy metodą K-Fold Cross Validation (przy K = 5), w każdym 'obrocie pętli' imputując brakujące wartości średnią z aktualnego zbioru treningowego (dla wartości 'max_depth' 3, 5, 10 oraz 15).

Porównaj wyniki obu metod, przy metodzie Cross Validation zwróć uwagę na odchylenia standardowe (metoda 'wyprodukuje' 5 wartości Accuracy dla każdego 'max_depth').

### Zadanie 2.4 (dodatkowe)
Powtórz kroki z zadania 2.3 dla pozostałych metod imputacji danych z zadania 1.1